Python/optimizer_analysis.c

```diff
@@ -547,38 +572,80 @@ peephole_opt(_PyInterpreterFrame *frame, _PyUOpInstruction *buffer, int buffer_s
                 buffer[pc].operand = (uintptr_t)val;
                 break;
             }
-            case _CHECK_PEP_523:
-            {
+            case _CHECK_PEP_523: {
                 /* Setting the eval frame function invalidates
                  * all executors, so no need to check dynamically */
                 if (_PyInterpreterState_GET()->eval_frame == NULL) {
                     buffer[pc].opcode = _NOP;
                 }
                 break;
             }
-            case _PUSH_FRAME:
-            case _POP_FRAME:
-            {
-                uint64_t operand = buffer[pc].operand;
-                if (operand & 1) {
-                    co = (PyCodeObject *)(operand & ~1);
-                    assert(PyCode_Check(co));
-                }
-                else if (operand == 0) {
-                    co = NULL;
+            case _CHECK_STACK_SPACE: {
+                assert(corresponding_check_stack == NULL);
+                corresponding_check_stack = &buffer[pc];
+                break;
+            }
+            case _PUSH_FRAME: {
+                assert(corresponding_check_stack != NULL);
+                co = get_co(&buffer[pc]);
+                if (co == NULL) {
+                    // should be about to _EXIT_TRACE anyway
+                    goto finish;
+                }
+                int framesize = co->co_framesize;
+                assert(framesize > 0);
+                curr_space += framesize;
+                if (curr_space < 0 || curr_space > INT32_MAX) {
```

Comment:
The second check cannot succeed when `int` is 32 bits, right? I guess the compiler will then omit the check.

Reply:
right. I was worried about cases where `int` is larger than the 32-bit operand. I'm comparing to `INT32_MAX` instead of `UINT32_MAX` to avoid compiler warnings; technically that makes this comparison 1 bit more strict than necessary, but I imagine this is an edge case that will almost never be hit anyway.

Reply:
🤔 actually, the comparison isn't overly strict, since I'm already requiring this number to be a positive int. Anyway, I'm still convinced this is a necessary check.